althea-mesh/clarity
# Automatic verification

This loads up Ethereum's tests for transactions, and dynamically creates a test case per JSON file.

This is WIP as there are ~19 tests which are dependent on network rules for verification. I.e. some are invalid on homestead, but can be valid on constantinople (i.e. since homestead valid sig assumes `0<s<secp256k1/n+1`). I've yet to figure out a nice way to implement it, but probably there should be some notion of "config" API to specify which network we're operating in.

There are also some transactions that are invalid on ALL networks, but there is something missing yet (i.e. `WrongVRSTestVEqual*`)

For simplicity I'll implement some of that per-network logic to be sure all tests can pass

This PR should close #14 and should close #8 too.

# How does it work

## Parse test cases

1. It reads all the test cases from Ethereum's test suite from a TransactionTests directory.
2. For each test fixture it also reads the filler data where expected values per  networks are found. For each set of network expectations (i.e. invalid on Homestead but invalid on Byzantium and Constantinople) it builds separate test case.

# Test case

1. Verify decoded RLP encoding, and compare with created `Transaction` based on filler data
2. Verify that encoding of built `Transaction` compares to RLP encoding of a test fixture.
3. Verify sender values where applicable (keep in mind that sometimes sender != expected sender and it is considered an expected failure; some test cases fails while decoding sender, so we're always trying to decode sender AND after that compare the string to the expected sender value).
4. Validate additional rules according to the expected protocol.

# Result

Test suite:

```
test result: ok. 198 passed; 0 failed; 0 ignored; 0 measured
```

Testnet verification:

First ever transaction made by clarity:

https://ropsten.etherscan.io/address/0x76af51f126e650fe26249e59d7d6d306203806f3
